File: src/main/java/recipeapplication/controller/PasswordController.java
package recipeapplication.controller;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.ResponseEntity;
import org.springframework.validation.Errors;
import org.springframework.web.bind.annotation.*;
import recipeapplication.dto.PasswordDto;
import recipeapplication.dto.UserDto;
import recipeapplication.exception.UserNotFoundException;
import recipeapplication.model.User;
import recipeapplication.service.AuthService;
import recipeapplication.service.UserService;

import javax.servlet.http.HttpServletRequest;
import javax.validation.Valid;

@RestController
@RequestMapping("/api")
public class PasswordController
{
    private UserService userService;
    private AuthService authService;

    @Autowired
    public PasswordController(UserService userService, AuthService authService)
    {
        this.userService = userService;
        this.authService = authService;
    }

    @PostMapping("/password_reset")
    public ResponseEntity createPasswordReset(HttpServletRequest request, @RequestBody UserDto userDto)
    {
        String serverName = String.format("%s:%s", request.getServerName(), request.getServerPort());

        try
        {
            userService.createPasswordResetToken(serverName, userDto);
        }
        catch (UserNotFoundException e)
        {
            // We don't show the client that the user doesn't exist for security reasons
        }

        return ResponseEntity.status(201).body("Created");
    }

    @PostMapping("/change_password")
    public ResponseEntity changePassword(@Valid @RequestBody PasswordDto passwordDto, Errors errors)
    {
        User user = authService.getLoggedInUser();

        if (user == null)
        {
            return ResponseEntity.status(400).body("Invalid session");
        }

        if (errors.hasErrors())
        {
            return ResponseEntity.status(400).body("Invalid password");
        }

        userService.changePassword(user, passwordDto.getPassword());

        return ResponseEntity.status(201).body("Created");
    }
}

File: src/main/kotlin/recipeapplication/controller/PasswordController.kt
package recipeapplication.controller

import org.springframework.beans.factory.annotation.Autowired
import org.springframework.http.ResponseEntity
import org.springframework.validation.Errors
import org.springframework.web.bind.annotation.PostMapping
import org.springframework.web.bind.annotation.RequestBody
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RestController
import recipeapplication.dto.PasswordDto
import recipeapplication.dto.PasswordResetDto
import recipeapplication.exception.UserNotFoundException
import recipeapplication.service.AuthService
import recipeapplication.service.UserService
import javax.servlet.http.HttpServletRequest
import javax.validation.Valid

@RestController
@RequestMapping("/api")
class PasswordController @Autowired constructor(private val userService: UserService, private val authService: AuthService) {
    @PostMapping("/password-reset")
    fun createPasswordReset(request: HttpServletRequest, @RequestBody passwordResetDto: PasswordResetDto): ResponseEntity<*> {
        val serverName = "${request.serverName}:${request.serverPort}"

        try {
            userService.createPasswordResetToken(serverName, passwordResetDto)
        } catch (e: UserNotFoundException) {
            // We don't show the client that the user doesn't exist for security reasons
        }

        return ResponseEntity.status(201).body("Created")
    }

    @PostMapping("/change-password")
    fun changePassword(@RequestBody passwordDto: @Valid PasswordDto, errors: Errors): ResponseEntity<*> {
        if (errors.hasErrors()) {
            return ResponseEntity.status(400).body("Invalid password")
        }

        userService.changePassword(authService.loggedInUser, passwordDto.password)

        return ResponseEntity.status(201).body("Created")
    }
}